### PR TITLE
Various improvements for CLI

### PR DESCRIPTION
**Pull request by aminya**
- Refactored the process of handling the missing options and configurations
- Made the configuration code more type-safe by using `Partial` and handling the missing properties.
- Calling `which` only if necessary
- Some other minor changes

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -12,33 +12,44 @@ const DEBUG_LOG = !!process.env.CMAKETSDEBUG;
...
   }
 
-  const configs = packJson['cmake-ts'] as BuildOptions;
+  const configs = packJson['cmake-ts'];
   if (!configs) {
     console.error('Package.json does not have cmake-ts key defined!');
     process.exit(1);
   }
 
-  const nativeonly = process.argv.some(arg => arg === 'nativeonly');
```
Comment (martin31821) [inline src/main.ts:30]: why change those?

File: src/main.ts [hunk 2]
```diff
@@ -49,10 +60,15 @@ const DEBUG_LOG = !!process.env.CMAKETSDEBUG;
...
-    console.log('--------------------------------------------------');
-    console.log('WARNING: Building only for the current OS.');
-    console.log('WARNING: DO NOT SHIP THE RESULTING PACKAGE');
-    console.log('--------------------------------------------------');
+    console.log(
+    `--------------------------------------------------
+      WARNING: Building only for the current OS.
+      WARNING: DO NOT SHIP THE RESULTING PACKAGE
+     --------------------------------------------------`);
+    if (configs.configurations === undefined) {
```
Comment (martin31821) [inline src/main.ts:68]: maybe this check should be moved to the top of the file?
i.e.exit immediately, if no build config could be found?

File: src/main.ts [hunk 3]
```diff
@@ -111,36 +131,37 @@ const DEBUG_LOG = !!process.env.CMAKETSDEBUG;
...
+      configs.generatorBinary = make;
     } else {
       console.error('Unsupported generator ' + configs.generatorToUse);
       process.exit(1);
     };
   }
 
-  const command = process.argv;
-  console.log('running in', packageDir, 'command', command);
+  console.log('running in', packageDir, 'command', /* command */ argv);
```
Comment (martin31821) [inline src/main.ts:164]: leftover /* command */ can be removed.

File: src/main.ts [hunk 4]
```diff
@@ -150,6 +171,10 @@ const DEBUG_LOG = !!process.env.CMAKETSDEBUG;
   await ensureDir(configs.stagingDirectory);
   console.log('[ DONE ]');
 
+  if (configs.configurations === undefined) {
```
Comment (martin31821) [inline src/main.ts:174]: duplicate from above?

**Review by martin31821 — CHANGES_REQUESTED** [review]
Looks good besides the two points I annotated to main.ts

**Comment by aminya** [comment]
@martin31821 Sorry I was still making the code pretty. 😄 I moved the whole defaulting process into separate functions to make the `main` function clean and understandable.

Could you re-review?

File: src/main.ts (hunk 2 above)
Comment (aminya) [reply src/main.ts:68]: Yes, I think we can handle this without error by using `defaultBuildConfigurations`. I believe that the tool should work even with minimal configurations.

File: src/main.ts (hunk 1 above)
Comment (aminya) [reply src/main.ts:30]: Yes, we can use `includes`. I don't think the performance is critical here as the array is small anyway.

Comment (aminya) [reply src/main.ts:30]: Fixed in 
https://github.com/EmbeddedEnterprises/cmake-ts/pull/12/commits/056eea42d04b54b84107844cb97e0fb2b53ed9db

File: src/main.ts (hunk 4 above)
Comment (aminya) [reply src/main.ts]: Fixed

File: src/main.ts (hunk 3 above)
Comment (aminya) [reply src/main.ts]: Fixed
https://github.com/EmbeddedEnterprises/cmake-ts/pull/12/commits/40b1b33ce71e3e43220bd5d5dcebff24e8a8683d

File: src/lib.ts [hunk 5]
```diff
@@ -32,6 +69,135 @@ export type BuildOptions = {
...
+    };
+  }
+
+  if (configs.buildType === undefined) {
+    configs.buildType = "Release";
+    console.warn("`buildType` was missing. Considering 'Release'");
+  }
+
+  // TODO move the code related to globalCMakeOptions
+  // TODO move the code related to customNANPackageName
```
Comment (martin31821) [inline src/lib.ts:195]: Do you want to move those or is that something to consider later?

File: src/main.ts [hunk 6]
```diff
@@ -1,146 +1,48 @@
...
 
 (async (): Promise<void> => {
 
-  const packageDir = process.cwd();
-  let packJson = null;
+  const argv = process.argv;
+  let packJson: {'cmake-ts': BuildOptions | undefined} & Record<string, any>;
   try {
-    packJson = require(resolve(join(packageDir, 'package.json')));
+    // TODO getting the path from the CLI
```
Comment (martin31821) [inline src/main.ts:17]: Fix in this PR or later?

**Review by martin31821 — COMMENTED** [review]
LGTM besides the two TODOs.

File: src/lib.ts (hunk 5 above)
Comment (aminya) [reply src/lib.ts:195]: Probably later. Maybe, we decide that it is not actually needed.

File: src/main.ts (hunk 6 above)
Comment (aminya) [reply src/main.ts:17]: Yes. This can be done later.

**Review by martin31821 — APPROVED** [review]

**Pull request merged**